File: appwin_notifications/android/src/main/kotlin/io/appwin/flutter/notifications/AppwinNotificationsPlugin.kt
package io.appwin.flutter.notifications

import android.content.Context
import android.os.Build
import io.appwin.core.AppwinCore
import io.appwin.core.availability.AppwinInitResult
import io.appwin.core.availability.AppwinInitStatus
import io.appwin.notifications.AppwinNotifications
import io.appwin.notifications.AutomationEvent
import io.appwin.notifications.InAppMessage
import io.appwin.notifications.TrackEvent
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch

/**
 * Dart-to-Kotlin glue for the Notifications product.
 *
 * Mirrors `AppwinNotificationsPlugin.swift`: same method names, same arguments,
 * same error codes. A divergence between the two only shows on one platform,
 * which is the worst way to find it.
 */
class AppwinNotificationsPlugin : FlutterPlugin, MethodChannel.MethodCallHandler {

  private var channel: MethodChannel? = null
  private var scope: CoroutineScope? = null
  private var appContext: Context? = null

  override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    appContext = binding.applicationContext
    scope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)
    channel = MethodChannel(binding.binaryMessenger, "appwin_notifications").also {
      it.setMethodCallHandler(this)
    }
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    channel?.setMethodCallHandler(null)
    channel = null
    scope?.cancel()
    scope = null
    appContext = null
  }

  override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
    when (call.method) {
      "getPlatformVersion" -> result.success("Android ${Build.VERSION.RELEASE}")

      "initialize" ->
        launch(result, "availability_failed") {
          result.success(encodeInitResult(AppwinNotifications.initialize()))
        }

      "registerPushToken" -> {
        val token = call.argument<String>("token")
        if (token.isNullOrBlank()) {
          result.error("bad_args", "token manquant", null)
          return
        }
        val optIn = call.argument<Boolean>("pushOptIn") ?: true
        launch(result, "register_failed") {
          AppwinCore.registerPushToken(token, pushOptIn = optIn)
          result.success(null)
        }
      }

      "trackEvent" -> {
        val raw = call.argument<String>("event")
        val event = AutomationEvent.entries.firstOrNull { it.wireValue == raw }
        if (event == null) {
          result.error("bad_args", "event manquant ou inconnu", null)
          return
        }
        val name = call.argument<String>("eventName")
        @Suppress("UNCHECKED_CAST")
        val properties = call.argument<Map<String, String>>("properties")
        launch(result, "track_event_failed") {
          AppwinNotifications.trackEvent(event, eventName = name, properties = properties)
          result.success(null)
        }
      }

      "fetchPendingMessages" ->
        launch(result, "fetch_failed") {
          result.success(AppwinNotifications.fetchPendingMessages().map(::encodeMessage))
        }

      "track" -> {
        val deliveryId = call.argument<String>("deliveryId")
        val raw = call.argument<String>("event")
        val event = TrackEvent.entries.firstOrNull { it.wireValue == raw }
        if (deliveryId.isNullOrBlank() || event == null) {
          result.error("bad_args", "deliveryId/event manquant", null)
          return
        }
        val buttonIndex = call.argument<Int>("buttonIndex")
        launch(result, "track_failed") {
          AppwinNotifications.track(deliveryId, event, buttonIndex)
          result.success(null)
        }
      }

      "syncOnAppOpen" ->
        launch(result, "sync_failed") {
          result.success(AppwinNotifications.syncOnAppOpen().map(::encodeMessage))
        }

      "start" -> {
        val ctx = appContext
        if (ctx == null) {
          result.error("detached", "plugin détaché du moteur Flutter", null)
          return
        }
        launch(result, "start_failed") {
          AppwinNotifications.start(ctx)
          result.success(null)
        }
      }

      "stop" -> {
        AppwinNotifications.stop()
        result.success(null)
      }

      "presentPendingMessages" ->
        launch(result, "present_failed") {
          AppwinNotifications.presentPendingMessages()
          result.success(null)
        }

      else -> result.notImplemented()
    }
  }

  private fun launch(result: MethodChannel.Result, code: String, block: suspend () -> Unit) {
    val current = scope
    if (current == null) {
      result.error("detached", "plugin détaché du moteur Flutter", null)
      return
    }
    current.launch {
      runCatching { block() }
        .onFailure { result.error(code, it.message, null) }
    }
  }
}

private fun encodeInitResult(result: AppwinInitResult): Map<String, Any> =
  when (result.status) {
    AppwinInitStatus.READY -> mapOf("status" to "ready")
    AppwinInitStatus.NOT_CONFIGURED -> mapOf("status" to "notConfigured")
    AppwinInitStatus.UNKNOWN -> mapOf("status" to "unknown")
    AppwinInitStatus.UNAVAILABLE ->
      mapOf("status" to "unavailable", "reason" to (result.reason?.key ?: "disabled"))
  }

private fun encodeMessage(message: InAppMessage): Map<String, Any?> {
  val content = mutableMapOf<String, Any?>(
    "title" to message.content.title,
    "body" to message.content.body,
    "imageUrl" to message.content.imageUrl,
    "deeplink" to message.content.deeplink,
  )
  message.content.buttons?.let { buttons ->
    content["buttons"] = buttons.map { button ->
      mapOf(
        "label" to button.label,
        "action" to button.action,
        "url" to button.url,
      )
    }
  }
  return mapOf(
    "id" to message.id,
    "campaignId" to message.campaignId,
    "deliveryId" to message.deliveryId,
    "channel" to message.channel,
    "format" to message.format,
    "content" to content,
  )
}

File: appwin_notifications/ios/appwin_notifications/Sources/appwin_notifications/AppwinNotificationsPlugin.swift
import Flutter
import UIKit
import AppwinCore
import AppwinNotifications

/**
 Dart-to-Swift glue for the Notifications product.

 Mirrors `AppwinNotificationsPlugin.kt`: same method names, same arguments,
 same error codes. A divergence between the two only shows on one platform,
 which is the worst way to find it.
 */
public class AppwinNotificationsPlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(
      name: "appwin_notifications",
      binaryMessenger: registrar.messenger()
    )
    registrar.addMethodCallDelegate(AppwinNotificationsPlugin(), channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "getPlatformVersion":
      result("iOS " + UIDevice.current.systemVersion)

    case "initialize":
      // Availability, not configuration: `AppwinCore.configure` is the host
      // app's job and runs through the appwin_core plugin.
      Task {
        let verdict = await AppwinNotifications.initialize()
        await MainActor.run { result(Self.encode(verdict)) }
      }

    case "registerPushToken":
      guard let args = call.arguments as? [String: Any],
            let token = args["token"] as? String
      else {
        result(Self.badArgs("token")); return
      }
      let optIn = args["pushOptIn"] as? Bool ?? true
      Task {
        do {
          try await AppwinCore.registerPushToken(token, pushOptIn: optIn)
          await MainActor.run { result(nil) }
        } catch {
          await Self.fail(result, "register_failed", error)
        }
      }

    case "trackEvent":
      guard let args = call.arguments as? [String: Any],
            let raw = args["event"] as? String,
            let event = AutomationEvent(rawValue: raw)
      else {
        result(Self.badArgs("event")); return
      }
      let name = args["eventName"] as? String
      let properties = args["properties"] as? [String: String]
      Task {
        do {
          try await AppwinNotifications.trackEvent(event, eventName: name, properties: properties)
          await MainActor.run { result(nil) }
        } catch {
          await Self.fail(result, "track_event_failed", error)
        }
      }

    case "fetchPendingMessages":
      Task {
        do {
          let messages = try await AppwinNotifications.fetchPendingMessages()
          await MainActor.run { result(messages.map(Self.encode)) }
        } catch {
          await Self.fail(result, "fetch_failed", error)
        }
      }

    case "track":
      guard let args = call.arguments as? [String: Any],
            let deliveryId = args["deliveryId"] as? String,
            let raw = args["event"] as? String,
            let event = TrackEvent(rawValue: raw)
      else {
        result(Self.badArgs("deliveryId/event")); return
      }
      let buttonIndex = args["buttonIndex"] as? Int
      Task {
        do {
          try await AppwinNotifications.track(
            deliveryId: deliveryId,
            event: event,
            buttonIndex: buttonIndex
          )
          await MainActor.run { result(nil) }
        } catch {
          await Self.fail(result, "track_failed", error)
        }
      }

    case "syncOnAppOpen":
      Task {
        do {
          let messages = try await AppwinNotifications.syncOnAppOpen()
          await MainActor.run { result(messages.map(Self.encode)) }
        } catch {
          await Self.fail(result, "sync_failed", error)
        }
      }

    case "start":
      let requestPush = (call.arguments as? [String: Any])?["requestPushPermission"] as? Bool ?? true
      Task {
        await AppwinNotifications.start(requestPushPermission: requestPush)
        await MainActor.run { result(nil) }
      }

    case "stop":
      Task { @MainActor in
        AppwinNotifications.stop()
        result(nil)
      }

    case "presentPendingMessages":
      Task {
        do {
          try await AppwinNotifications.presentPendingMessages()
          await MainActor.run { result(nil) }
        } catch {
          await Self.fail(result, "present_failed", error)
        }
      }

    default:
      result(FlutterMethodNotImplemented)
    }
  }

  // MARK: - Encoding

  /// Maps the native result onto what the Dart side parses.
  ///
  /// A dictionary rather than a raw string: the reason travels with the
  /// status, and the two must not drift apart across the channel.
  static func encode(_ result: AppwinInitResult) -> [String: Any] {
    switch result {
    case .ready:
      return ["status": "ready"]
    case .notConfigured:
      return ["status": "notConfigured"]
    case .unknown:
      return ["status": "unknown"]
    case .unavailable(let reason):
      return ["status": "unavailable", "reason": reason.rawValue]
    }
  }

  static func encode(_ message: InAppMessage) -> [String: Any] {
    var content: [String: Any] = [:]
    if let title = message.content.title { content["title"] = title }
    if let body = message.content.body { content["body"] = body }
    if let imageUrl = message.content.imageUrl { content["imageUrl"] = imageUrl }
    if let deeplink = message.content.deeplink { content["deeplink"] = deeplink }
    if let buttons = message.content.buttons {
      content["buttons"] = buttons.map { button in
        var map: [String: Any] = ["label": button.label, "action": button.action.rawValue]
        if let url = button.url { map["url"] = url }
        return map
      }
    }
    return [
      "id": message.id,
      "campaignId": message.campaignId,
      "deliveryId": message.deliveryId,
      "channel": message.channel.rawValue,
      "format": message.format.rawValue,
      "content": content,
    ]
  }

  static func badArgs(_ missing: String) -> FlutterError {
    FlutterError(code: "bad_args", message: "\(missing) manquant", details: nil)
  }

  @MainActor
  static func fail(_ result: @escaping FlutterResult, _ code: String, _ error: Error) {
    result(FlutterError(code: code, message: error.localizedDescription, details: nil))
  }
}
